### PR TITLE
add thinning support to fori_collect

### PR DESCRIPTION
**Pull request by martinjankowiak**

### CODE REVIEW

File: numpyro/util.py [hunk 1]
```diff
@@ -197,26 +200,28 @@ def fori_collect(lower, upper, body_fun, init_val, transform=identity,
         collected along the leading axis of `np.ndarray` objects.
     """
     assert lower <= upper
-    collection_size = upper - lower if collection_size is None else collection_size
-    assert collection_size >= upper - lower
+    assert thinning >= 0
+    collection_size = (upper - lower) // thinning if collection_size is None else collection_size // thinning
```
Comment (martinjankowiak) [inline numpyro/util.py:204]: @fehiepsi not quite sure what the interaction should be when `collection_size is not None`

File: numpyro/util.py [hunk 2]
```diff
@@ -197,26 +200,28 @@ def fori_collect(lower, upper, body_fun, init_val, transform=identity,
...
     if not progbar:
-        last_val, collection, _ = fori_loop(0, upper, _body_fn, (init_val, collection, lower))
+        last_val, collection, _, _ = fori_loop(0, upper, _body_fn, (init_val, collection, start_idx, thinning))
     else:
         diagnostics_fn = progbar_opts.pop('diagnostics_fn', None)
         progbar_desc = progbar_opts.pop('progbar_desc', lambda x: '')
 
-        vals = (init_val, collection, device_put(lower))
+        vals = (init_val, collection, device_put(start_idx), device_put(thinning))
         if upper == 0:
```
Comment (martinjankowiak) [inline numpyro/util.py:225]: also not really sure if this special case needs modification

File: numpyro/infer/mcmc.py [hunk 3]
```diff
@@ -262,7 +264,7 @@ def __init__(self,
         self._init_state_cache = {}
         self._cache = {}
         self._collection_params = {}
-        self._set_collection_params()
+        self._set_collection_params(thinning=self.thinning)
```
Comment (martinjankowiak) [inline numpyro/infer/mcmc.py:267]: so this works but i'm not really clear on how these args are piped through...

File: numpyro/util.py [hunk 4]
```diff
@@ -197,26 +200,28 @@ def fori_collect(lower, upper, body_fun, init_val, transform=identity,
...
     @cached_by(fori_collect, body_fun, transform)
     def _body_fn(i, vals):
-        val, collection, lower_idx = vals
+        val, collection, start_idx, thinning = vals
         val = body_fun(val)
-        i = jnp.where(i >= lower_idx, i - lower_idx, 0)
-        collection = ops.index_update(collection, i, ravel_pytree(transform(val))[0])
-        return val, collection, lower_idx
+        idx = (i - start_idx) // thinning
+        collection = ops.index_update(collection, idx, ravel_pytree(transform(val))[0])
```
Comment (fehiepsi) [inline numpyro/util.py:214]: nit: how about setting
```python
collection = cond(idx >= 0,
                  collection,
                  lambda x: ops.index_update(x, idx, ravel_pytree(transform(val))[0]),
                  collection,
                  identity)
```
to only update collection when the `idx` is nonnegative.

**Review by fehiepsi — COMMENTED** [review]
Thanks, @martinjankowiak! LGTM modulo a few nits.

File: numpyro/util.py [hunk 5]
```diff
@@ -197,26 +200,28 @@ def fori_collect(lower, upper, body_fun, init_val, transform=identity,
...
+    assert thinning >= 0
+    collection_size = (upper - lower) // thinning if collection_size is None else collection_size // thinning
+    assert collection_size >= (upper - lower) // thinning
     init_val_flat, unravel_fn = ravel_pytree(transform(init_val))
+    start_idx = lower + (upper - lower) % thinning
 
     @cached_by(fori_collect, body_fun, transform)
     def _body_fn(i, vals):
-        val, collection, lower_idx = vals
+        val, collection, start_idx, thinning = vals
```
[the remark this answers is not in the document]
Comment (martinjankowiak) [reply numpyro/util.py:211]: are you sure i don't think it worked otherwise

[the remark this answers is not in the document]
Comment (fehiepsi) [reply numpyro/util.py:211]: Hmm... I am not sure. Probably when `start_idx or thinning` changed, we still use the previous `_body_fn` here due to `cached_by` logic. Thanks for pointing it out. Leave it as-is is safer.

File: numpyro/infer/mcmc.py [hunk 6]
```diff
@@ -312,15 +319,17 @@ def _single_chain_mcmc(self, init, args, kwargs, collect_fields):
         lower_idx = self._collection_params["lower"]
         upper_idx = self._collection_params["upper"]
         phase = self._collection_params["phase"]
-
+        collection_size = self._collection_params["collection_size"]
+        collection_size = collection_size if collection_size is None else collection_size // self.thinning
```
Comment (martinjankowiak) [inline numpyro/infer/mcmc.py:323]: is this what you meant?

Comment (fehiepsi) [reply numpyro/infer/mcmc.py:323]: yes, it is better than what i meant

File: numpyro/util.py [hunk 7]
```diff
@@ -185,6 +186,8 @@ def fori_collect(lower, upper, body_fun, init_val, transform=identity,
     :param progbar: whether to post progress bar updates.
     :param bool return_last_val: If `True`, the last value is also returned.
         This has the same type as `init_val`.
+    :param thinning: Positive integer that controls the thinning ratio for retained
+        values. Defaults to 1, i.e. no thinning.
     :param int collection_size: Size of the returned collection. If not specified,
         the size will be ``upper - lower``. If the size is larger than
         ``upper - lower``, only the top ``upper - lower`` entries will be non-zero.
```
Comment (fehiepsi) [inline numpyro/util.py:193]: All ``upper - lower`` -> ``(upper - lower) // thinning``

**Review by fehiepsi — APPROVED** [review]
LGTM, thanks @martinjankowiak! Just have a comment regarding the docs of collection_size

File: numpyro/util.py [hunk 8]
```diff
@@ -197,26 +200,32 @@ def fori_collect(lower, upper, body_fun, init_val, transform=identity,
         collected along the leading axis of `np.ndarray` objects.
     """
     assert lower <= upper
-    collection_size = upper - lower if collection_size is None else collection_size
-    assert collection_size >= upper - lower
+    assert thinning >= 1 and thinning <= upper - lower
```
Comment (fehiepsi) [inline numpyro/util.py:203]: You can relax the condition `thinning <= upper - lower`. It is not needed.

**Review by fehiepsi — APPROVED** [review]

**Pull request merged**